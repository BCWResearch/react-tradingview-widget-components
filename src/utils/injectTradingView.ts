declare global {
    interface Window {
        TradingView?: TradingView;
    }
}

type Widget = { new (options: Record<string, unknown>): {id: string; options: Record<string, unknown>;} }

export interface TradingView {
    host: string;
    ideasHost: string;
    chatHost: string;
    widgetHost: string;
    getHost(e: unknown): string;
    embedStylesForCopyright(): void;
    embedStylesForFullHeight(n: number, hasCopyright: boolean, id: string): void;
    gId(): string;
    isPersentHeight(): boolean;
    getSuffix(): string;
    hasCopyright(id: string): boolean;
    calculateWidgetHeight(num: number, element: unknown): string;
    onready(e: Event): void;
    css(e: unknown): void;
    bindEvent(e: unknown, t: unknown, o: unknown): void;
    unbindEvent(e: unknown, t: unknown, o: unknown): void;
    cloneSimpleObject(e: unknown): typeof e;
    isArray(e: unknown): boolean;
    isMobileDevice(device?: Record<string, boolean>): boolean;
    generateUtmForUrlParams(e: unknown): string;
    WidgetAbstract(): Record<string, unknown>,
    MiniWidget: Widget
    MediumWidget: Widget;
    widget: Widget;
    chart: Widget;
    stream: Widget;
    EventsWidget: Widget;
    IdeasStreamWidget: Widget;
    IdeaWidget: Widget;
    ChatWidgetEmbed: Widget;
}

export const injectTradingView = (signal?: AbortSignal) => {
	return new Promise<TradingView>((resolve, reject) => {
		if (window.TradingView) {
			resolve(window.TradingView);
			return;
		}
		const script = document.createElement("script");
		script.type = "text/javascript";
		script.src = "https://s3.tradingview.com/tv.js";
		script.async = true;
		script.onload = () => {
			if (signal?.aborted) {
				reject(new Error("TradingView Widget script load aborted."));
				return;
			}
			if (!window.TradingView) {
				reject(new Error("Failed to inject TradingView into window"));
				return;
			}
			resolve(window.TradingView);
		};
		script.onerror = () => reject("Failed to inject Trading View script");
		if (!document.querySelector("script[src='https://s3.tradingview.com/tv.js']")) document.head.appendChild(script);
	});
};