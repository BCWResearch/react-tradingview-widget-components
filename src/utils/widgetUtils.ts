export const getId = (): string => {
	return "tradingview_" + (1048576 * (1 + Math.random()) | 0).toString(16).substring(1);
};