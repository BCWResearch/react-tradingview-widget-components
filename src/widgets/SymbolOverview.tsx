export const SymbolOverview = () => {
	return null;
};