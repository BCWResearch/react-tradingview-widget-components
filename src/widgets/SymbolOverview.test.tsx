import {
	render, screen 
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { SymbolOverview } from "./SymbolOverview";
import { injectTradingView } from "../utils/injectTradingView";

const mockedInjectTradingView = jest.mocked(injectTradingView);

jest.mock("../utils/injectTradingView", () => ({
	injectTradingView: jest.fn(async () => ({
		MediumWidget: function(props: unknown) {
			const node = document.querySelector("#mockedId");
			if (node) node.innerHTML = "WIDGET";
			return props;
		} 
	})), 
}));

jest.mock("../utils/widgetUtils", () => ({
	getId: jest.fn(() => "mockedId")
}));

test("renders widget inside div with id", async () => {
	render(<SymbolOverview/>);
	expect(await screen.findByText(/WIDGET/)).toBeInTheDocument();
});

test("containing div id should be the same on rerender", async () => {
	const { rerender, container } = render(<SymbolOverview/>);
	expect(await screen.findByText(/WIDGET/)).toBeInTheDocument();
	expect(container).toMatchSnapshot();
	rerender(<SymbolOverview chartOnly/>);
	expect(container).toMatchSnapshot();
});

test("should render fallback on error", async () => {
	mockedInjectTradingView.mockImplementation(async () => {throw new Error();});
	render(<SymbolOverview fallback="ERROR"/>);
	expect(await screen.findByText(/ERROR/)).toBeInTheDocument();
});

