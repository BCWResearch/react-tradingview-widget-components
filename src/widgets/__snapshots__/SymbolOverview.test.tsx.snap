// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containing div id should be the same on rerender 1`] = `
<div>
  <div>
    <div
      id="mockedId"
    >
      WIDGET
    </div>
  </div>
</div>
`;

exports[`containing div id should be the same on rerender 2`] = `
<div>
  <div>
    <div
      id="mockedId"
    >
      WIDGET
    </div>
  </div>
</div>
`;
